### ISSUE TITLE
Unity iOS IL2CPP Build error

### ISSUE THREAD

**Issue by housei** [msgpack/msgpack-cli#101]
OS: OSX 10.10.3
Unity:5.1.2p1
use:MessagePack Cli 0.6.0-beta5 unity3d/MsgPack.dll

Unity iOS IL2CPP Build error I will report since occurred.
Workaround I will want you to tell me if any.

-----
IL2CPP error for method 'TAction MsgPack.Serialization.NilImplicationHandler`4::OnUnpacked(TUnpackedParameter,MsgPack.Serialization.NilImplication)' in assembly '/Users/xxxxxx/Documents/unity_workspace/xxxxx-unity/Temp/StagingArea/Data/Managed/Mono.Security.dll'
Additional information: Build a development build for more information. The given key was not present in the dictionary.

Failed running /Applications/Unity5.1.2p1/Unity.app/Contents/Frameworks/il2cpp/build/il2cpp.exe --copy-level=None --emit-null-checks --enable-array-bounds-check --enable-symbol-loading --development-mode --extra-types.file="/Applications/Unity5.1.2p1/Unity.app/Contents/Frameworks/il2cpp/il2cpp_default_extra_types.txt"  --assembly="/Users/xxxxxx/Documents/unity_workspace/xxxxx-unity/Temp/StagingArea/Data/Managed/Assembly-CSharp.dll" --assembly="/Users/xxxxxx/Documents/unity_workspace/xxxxx-unity/Temp/StagingArea/Data/Managed/UnityEngine.UI.dll" --assembly="/Users/xxxxxx/Documents/unity_workspace/xxxxx-unity/Temp/StagingArea/Data/Managed/UnityEngine.dll" --generatedcppdir="/Users/xxxxxx/Documents/unity_workspace/xxxxx-unity/Temp/il2cppOutput/il2cppOutput" --builder=none 

stdout:
IL2CPP error for method 'TAction MsgPack.Serialization.NilImplicationHandler`4::OnUnpacked(TUnpackedParameter,MsgPack.Serialization.NilImplication)' in assembly '/Users/xxxxxx/Documents/unity_workspace/xxxxx-unity/Temp/StagingArea/Data/Managed/Mono.Security.dll'
Additional information: Build a development build for more information. The given key was not present in the dictionary.
il2cpp.exe didn't catch exception: System.Collections.Generic.KeyNotFoundException: The given key was not present in the dictionary.
  at System.Collections.Generic.Dictionary`2[Mono.Cecil.TypeReference,System.Int32].get_Item (Mono.Cecil.TypeReference key) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.Metadata.VTableBuilder.GetVirtualMethodTargetMethod (Mono.Cecil.TypeReference type, Mono.Cecil.MethodReference method) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodBodyWriter.ArgsFor (Mono.Cecil.MethodReference method, Mono.Cecil.MethodReference unresolvedMethod, MethodCallType callType, System.Collections.Generic.Stack`1 valueStack, Unity.IL2CPP.ILPreProcessor.TypeResolver typeResolverForCallingMethod, Unity.IL2CPP.ILPreProcessor.TypeResolver typeResolverForMethodToCall, IRuntimeMetadataAccess runtimeMetadata, SharingType sharingType, Unity.IL2CPP.Metadata.VTableBuilder vTableBuilder, Boolean isConstructor, Mono.Cecil.TypeReference constrainedType) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodBodyWriter.CallMethod (Mono.Cecil.MethodReference callingMethod, Mono.Cecil.MethodReference methodToCall, MethodCallType callType, Mono.Cecil.Cil.Instruction instruction, Mono.Cecil.TypeReference constrainedType) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodBodyWriter.ProcessInstruction (Unity.IL2CPP.Node node, Unity.IL2CPP.Common.CFG.InstructionBlock block, Mono.Cecil.Cil.Instruction& ins) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodBodyWriter.GenerateCodeRecursive (Unity.IL2CPP.Node node) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodBodyWriter.Generate () [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodWriter.WriteMethodBody (Mono.Cecil.MethodReference method, Unity.IL2CPP.CppCodeWriter methodBodyWriter, IRuntimeMetadataAccess metadataAccess, Unity.IL2CPP.IncludeCollection.IncludeCollection includeCollection) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodWriter+<WriteMethodDefinition>c__AnonStorey0.<>m__0 (Unity.IL2CPP.CppCodeWriter bodyWriter, Unity.IL2CPP.MetadataUsage metadataUsage, Unity.IL2CPP.MethodUsage methodUsage) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodWriter.WriteMethodWithMetadataInitialization (Unity.IL2CPP.CppCodeWriter writer, System.String methodSignature, System.Action`3 writeMethodBody) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodWriter.WriteMethodDefinition (Mono.Cecil.MethodReference method, Unity.IL2CPP.MethodCollector methodCollector, Unity.IL2CPP.IncludeCollection.IncludeCollection includeCollection) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodWriter.WriteMethodDefinitions (System.Func`2 filter, Unity.IL2CPP.MethodCollector methodCollector, Unity.IL2CPP.IncludeCollection.IncludeCollection includeCollection) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.SourceWriter.WriteMethodSourceFiles (NiceIO.NPath outputDirectory, System.String fileName, IEnumerable`1 typeList, Unity.IL2CPP.MethodCollector methodCollector) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.SourceWriter.WriteGenerics (Unity.IL2CPP.GenericsCollection.InflatedCollection genericsCollection, IEnumerable`1 allTypeDefinitions, Unity.IL2CPP.MethodCollector methodCollector) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.AssemblyConverter.WriteGenerics (Unity.IL2CPP.GenericsCollection.InflatedCollection allGenerics, IEnumerable`1 allTypeDefinitions) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.AssemblyConverter.Apply () [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.AssemblyConverter.ConvertAssemblies (System.String[] assemblies, NiceIO.NPath outputDir, System.Collections.Generic.List`1 additionalCpp) [0x00000] in <filename unknown>:0 
stderr:

Unhandled Exception:
System.Collections.Generic.KeyNotFoundException: The given key was not present in the dictionary.
  at System.Collections.Generic.Dictionary`2[Mono.Cecil.TypeReference,System.Int32].get_Item (Mono.Cecil.TypeReference key) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.Metadata.VTableBuilder.GetVirtualMethodTargetMethod (Mono.Cecil.TypeReference type, Mono.Cecil.MethodReference method) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodBodyWriter.ArgsFor (Mono.Cecil.MethodReference method, Mono.Cecil.MethodReference unresolvedMethod, MethodCallType callType, System.Collections.Generic.Stack`1 valueStack, Unity.IL2CPP.ILPreProcessor.TypeResolver typeResolverForCallingMethod, Unity.IL2CPP.ILPreProcessor.TypeResolver typeResolverForMethodToCall, IRuntimeMetadataAccess runtimeMetadata, SharingType sharingType, Unity.IL2CPP.Metadata.VTableBuilder vTableBuilder, Boolean isConstructor, Mono.Cecil.TypeReference constrainedType) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodBodyWriter.CallMethod (Mono.Cecil.MethodReference callingMethod, Mono.Cecil.MethodReference methodToCall, MethodCallType callType, Mono.Cecil.Cil.Instruction instruction, Mono.Cecil.TypeReference constrainedType) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodBodyWriter.ProcessInstruction (Unity.IL2CPP.Node node, Unity.IL2CPP.Common.CFG.InstructionBlock block, Mono.Cecil.Cil.Instruction& ins) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodBodyWriter.GenerateCodeRecursive (Unity.IL2CPP.Node node) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodBodyWriter.Generate () [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodWriter.WriteMethodBody (Mono.Cecil.MethodReference method, Unity.IL2CPP.CppCodeWriter methodBodyWriter, IRuntimeMetadataAccess metadataAccess, Unity.IL2CPP.IncludeCollection.IncludeCollection includeCollection) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodWriter+<WriteMethodDefinition>c__AnonStorey0.<>m__0 (Unity.IL2CPP.CppCodeWriter bodyWriter, Unity.IL2CPP.MetadataUsage metadataUsage, Unity.IL2CPP.MethodUsage methodUsage) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodWriter.WriteMethodWithMetadataInitialization (Unity.IL2CPP.CppCodeWriter writer, System.String methodSignature, System.Action`3 writeMethodBody) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodWriter.WriteMethodDefinition (Mono.Cecil.MethodReference method, Unity.IL2CPP.MethodCollector methodCollector, Unity.IL2CPP.IncludeCollection.IncludeCollection includeCollection) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodWriter.WriteMethodDefinitions (System.Func`2 filter, Unity.IL2CPP.MethodCollector methodCollector, Unity.IL2CPP.IncludeCollection.IncludeCollection includeCollection) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.SourceWriter.WriteMethodSourceFiles (NiceIO.NPath outputDirectory, System.String fileName, IEnumerable`1 typeList, Unity.IL2CPP.MethodCollector methodCollector) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.SourceWriter.WriteGenerics (Unity.IL2CPP.GenericsCollection.InflatedCollection genericsCollection, IEnumerable`1 allTypeDefinitions, Unity.IL2CPP.MethodCollector methodCollector) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.AssemblyConverter.WriteGenerics (Unity.IL2CPP.GenericsCollection.InflatedCollection allGenerics, IEnumerable`1 allTypeDefinitions) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.AssemblyConverter.Apply () [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.AssemblyConverter.ConvertAssemblies (System.String[] assemblies, NiceIO.NPath outputDir, System.Collections.Generic.List`1 additionalCpp) [0x00000] in <filename unknown>:0 
[ERROR] FATAL UNHANDLED EXCEPTION: System.Collections.Generic.KeyNotFoundException: The given key was not present in the dictionary.
  at System.Collections.Generic.Dictionary`2[Mono.Cecil.TypeReference,System.Int32].get_Item (Mono.Cecil.TypeReference key) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.Metadata.VTableBuilder.GetVirtualMethodTargetMethod (Mono.Cecil.TypeReference type, Mono.Cecil.MethodReference method) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodBodyWriter.ArgsFor (Mono.Cecil.MethodReference method, Mono.Cecil.MethodReference unresolvedMethod, MethodCallType callType, System.Collections.Generic.Stack`1 valueStack, Unity.IL2CPP.ILPreProcessor.TypeResolver typeResolverForCallingMethod, Unity.IL2CPP.ILPreProcessor.TypeResolver typeResolverForMethodToCall, IRuntimeMetadataAccess runtimeMetadata, SharingType sharingType, Unity.IL2CPP.Metadata.VTableBuilder vTableBuilder, Boolean isConstructor, Mono.Cecil.TypeReference constrainedType) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodBodyWriter.CallMethod (Mono.Cecil.MethodReference callingMethod, Mono.Cecil.MethodReference methodToCall, MethodCallType callType, Mono.Cecil.Cil.Instruction instruction, Mono.Cecil.TypeReference constrainedType) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodBodyWriter.ProcessInstruction (Unity.IL2CPP.Node node, Unity.IL2CPP.Common.CFG.InstructionBlock block, Mono.Cecil.Cil.Instruction& ins) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodBodyWriter.GenerateCodeRecursive (Unity.IL2CPP.Node node) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodBodyWriter.Generate () [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodWriter.WriteMethodBody (Mono.Cecil.MethodReference method, Unity.IL2CPP.CppCodeWriter methodBodyWriter, IRuntimeMetadataAccess metadataAccess, Unity.IL2CPP.IncludeCollection.IncludeCollection includeCollection) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodWriter+<WriteMethodDefinition>c__AnonStorey0.<>m__0 (Unity.IL2CPP.CppCodeWriter bodyWriter, Unity.IL2CPP.MetadataUsage metadataUsage, Unity.IL2CPP.MethodUsage methodUsage) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodWriter.WriteMethodWithMetadataInitialization (Unity.IL2CPP.CppCodeWriter writer, System.String methodSignature, System.Action`3 writeMethodBody) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodWriter.WriteMethodDefinition (Mono.Cecil.MethodReference method, Unity.IL2CPP.MethodCollector methodCollector, Unity.IL2CPP.IncludeCollection.IncludeCollection includeCollection) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.MethodWriter.WriteMethodDefinitions (System.Func`2 filter, Unity.IL2CPP.MethodCollector methodCollector, Unity.IL2CPP.IncludeCollection.IncludeCollection includeCollection) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.SourceWriter.WriteMethodSourceFiles (NiceIO.NPath outputDirectory, System.String fileName, IEnumerable`1 typeList, Unity.IL2CPP.MethodCollector methodCollector) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.SourceWriter.WriteGenerics (Unity.IL2CPP.GenericsCollection.InflatedCollection genericsCollection, IEnumerable`1 allTypeDefinitions, Unity.IL2CPP.MethodCollector methodCollector) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.AssemblyConverter.WriteGenerics (Unity.IL2CPP.GenericsCollection.InflatedCollection allGenerics, IEnumerable`1 allTypeDefinitions) [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.AssemblyConverter.Apply () [0x00000] in <filename unknown>:0 
  at Unity.IL2CPP.AssemblyConverter.ConvertAssemblies (System.String[] assemblies, NiceIO.NPath outputDir, System.Collections.Generic.List`1 additionalCpp) [0x00000] in <filename unknown>:0 

UnityEngine.Debug:LogError(Object)
UnityEditorInternal.Runner:RunManagedProgram(String, String, String, CompilerOutputParserBase) (at /Users/builduser/buildslave/unity/build/Editor/Mono/BuildPipeline/Il2Cpp/IL2CPPUtils.cs:66)
UnityEditorInternal.IL2CPPBuilder:ConvertPlayerDlltoCpp(ICollection`1, String, String) (at /Users/builduser/buildslave/unity/build/Editor/Mono/BuildPipeline/Il2Cpp/IL2CPPUtils.cs:402)
UnityEditorInternal.IL2CPPBuilder:Run() (at /Users/builduser/buildslave/unity/build/Editor/Mono/BuildPipeline/Il2Cpp/IL2CPPUtils.cs:293)
UnityEditorInternal.IL2CPPUtils:RunIl2Cpp(String, String, IIl2CppPlatformProvider, Action`1, RuntimeClassRegistry) (at /Users/builduser/buildslave/unity/build/Editor/Mono/BuildPipeline/Il2Cpp/IL2CPPUtils.cs:211)
UnityEditor.HostView:OnGUI()

**Comment by neuecc** [comment]
:+1: 
I've encounted same error.

**Comment by neuecc** [comment]
It can reproduce at Windows environment.

Add MsgPack.dll and project settings to iOS(Debug/Develpment Build/Script Debugging) + IL2CPP.
and put following code.

```csharp
// with MsgPack-CLI

using UnityEngine;

public class NewBehaviourScript : MonoBehaviour
{
    void Start()
    {
        var _ = MsgPack.Serialization.MessagePackSerializer.Get<int>().UnpackSingleObject(null);
        UnityEngine.Debug.Log(_);
    }
}
```

Maybe it is caused by MsgPack.Serialization.NilImplicationHandler`4::OnUnpacked.
It is maybe Unity's IL2CPP compiler's bug.
I've reported to Unity(but I can't create small reproduce code)

I do not know when Unity revises it.
If you can find IL2CPP compiler's bug, I want to you fix avoid the problem.

**Comment by yfakariya** [comment]
@neuecc  Thank you for reporting. I found minimal repro code as following:

```c#
using UnityEngine;

public class NewBehaviourScript : MonoBehaviour {

	void Start () {
		// This line causes IL2CPP Error!
		UnityEngine.Debug.Log(new Generic<string, ImplementingValueType>().Foo(new ImplementingValueType()));
	}

	void Update () {
	
	}
}

class Generic<T, TFactory>
	where TFactory : IInterface<T> // Point1 : The generic parameter is contrained by another generic parameter.
{
	public T Foo(TFactory generic)
	{
		return generic.Bar(); // Point2: Invoke method via generic parameter constraint.
	}
}

interface IInterface<T>
{
	T Bar();
}

// Point3: Actual generic argument, which is constrained as some interface, is value type instead of reference type.
struct ImplementingValueType : IInterface<string>
{
	public string Bar()
	{
		return "Hello";
	}
}
```
It looks a bug of combination of value type and generic constraint.

**Comment by yfakariya** [comment]
I released fixed version 0.6.1.1 now. Please try it. https://github.com/msgpack/msgpack-cli/releases/tag/0.6.1.1
